lib/Parse/ParseExpr.cpp

```diff
@@ -2115,95 +2197,26 @@ DeclNameRef Parser::parseUnqualifiedDeclBaseName(
-                                             bool allowOperators,
-                                             bool allowZeroArgCompoundNames,
-                                             bool allowDeinitAndSubscript) {
-  // Consume the base name.
-  auto baseName = parseUnqualifiedDeclBaseName(afterDot, loc, diag,
-                                               allowOperators,
-                                               allowDeinitAndSubscript);
-
-  // If the next token isn't a following '(', we don't have a compound name.
-  if (!baseName || !Tok.isFollowingLParen())
-    return baseName;
-
-  // If the next token is a ')' then we have a 0-arg compound name. This is
-  // explicitly differentiated from "simple" (non-compound) name in DeclName.
-  // Unfortunately only some places in the grammar are ok with accepting this
-  // kind of name; in other places it's ambiguous with trailing calls.
-  if (allowZeroArgCompoundNames && peekToken().is(tok::r_paren)) {
-    SyntaxParsingContext ArgsCtxt(SyntaxContext, SyntaxKind::DeclNameArguments);
-    consumeToken(tok::l_paren);
-    if (SyntaxContext->isEnabled())
-      SyntaxContext->addSyntax(
-          ParsedSyntaxRecorder::makeBlankDeclNameArgumentList(
-              leadingTriviaLoc(), *SyntaxContext));
-    consumeToken(tok::r_paren);
-    SmallVector<Identifier, 2> argumentLabels;
-    return baseName.withArgumentLabels(Context, argumentLabels);
-  }
-
-  // If the token after that isn't an argument label or ':', we don't have a
-  // compound name.
-  if ((!peekToken().canBeArgumentLabel() && !peekToken().is(tok::colon)) ||
-      Identifier::isEditorPlaceholder(peekToken().getText())) {
-    return baseName;
-  }
-
-  // Try to parse a compound name.
-  SyntaxParsingContext ArgsCtxt(SyntaxContext, SyntaxKind::DeclNameArguments);
-  BacktrackingScope backtrack(*this);
-
+  // Parse an argument list, if the flags allow it and it's present.
   SmallVector<Identifier, 2> argumentLabels;
   SmallVector<SourceLoc, 2> argumentLabelLocs;
-  SourceLoc lparenLoc = consumeToken(tok::l_paren);
+  SourceLoc lparenLoc;
   SourceLoc rparenLoc;
-  while (Tok.isNot(tok::r_paren)) {
-    SyntaxParsingContext ArgCtxt(SyntaxContext, SyntaxKind::DeclNameArgument);
-
-    // If we see a ':', the user forgot the '_';
-    if (Tok.is(tok::colon)) {
-      diagnose(Tok, diag::empty_arg_label_underscore)
-          .fixItInsert(Tok.getLoc(), "_");
-      argumentLabels.push_back(Identifier());
-      argumentLabelLocs.push_back(consumeToken(tok::colon));
-    }
 
-    Identifier argName;
-    SourceLoc argLoc;
-    parseOptionalArgumentLabel(argName, argLoc);
-    if (argLoc.isValid()) {
-      argumentLabels.push_back(argName);
-      argumentLabelLocs.push_back(argLoc);
-      continue;
-    }
+  bool hadArgList = tryParseArgLabelList(*this, flags, lparenLoc,
+                                         argumentLabels, argumentLabelLocs,
+                                         rparenLoc);
```

Comment:
This unconditionally tries to parse argument labels part. Don't we need to differentiate former `parseUnqualifiedDeclBaseName()` call?
For example, for `Foo(foo:bar:)` at type position, we should not parse it as valid `IdentTypeRepr`, but should stop parsing before the argument part (and diagnose it).

Reply:
The first thing `tryParseArgLabelList` does is check if the `AllowCompoundNames` flag is set and bail out if it isn’t. I could hoist that check into this method, though.

Reply:
Ah I see. Thanks for explanation!